mlc_llm/relax_model/param_manager.py

```diff
@@ -215,40 +362,171 @@ def quantization_transform(
-            
 
         # Create the function mutator for applying dequantization.
         replacer = ParamReplacer(mod, func2param_var, f_replace)
         # Update the input IRModule with dequantization.
         mod = replacer.transform()
 
+        ######################## Part 4. Merge IRModule ########################
+
         # Merge the quantization IRModule into the updated input IRModule.
         for gv, func in mod_transform.functions.items():
             mod[gv] = func
+        return mod
+
+    def get_param_loading_functions(
+        self,
+        model_params: List[Optional[tvm.nd.NDArray]],
+        loaded_params: List[tvm.nd.NDArray],
+        loaded_idx_set: Set[int],
+        loaded_torch_bins: Set[str],
+        cached_relax_params: Dict[int, tvm.nd.NDArray],
+        cached_torch_params: Dict[str, Any],
+        device: Device,
+        device_cpu: Device,
+    ) -> Tuple[Callable, Callable]:
+        """A wrapper function which returns the `get_item` and `set_item`
+        functions for parameter lazy loading.
+
+        Parameters
+        ----------
+        model_params : List[Optional[tvm.nd.NDArray]]
+            The pre-loaded model parameters, for which we skip lazy loading.
 
-        # Return the merged IRModule and the mapping from each parameter's
-        # index to the parameter' name.
-        return mod, {pidx: pname for pidx, pname in enumerate(self.param_names)}
+        loaded_params : List[tvm.nd.NDArray]
+            The parameter loading result, storing all the loaded parameters.
+
+        loaded_idx_set : Set[int]
+            The set of indices of loaded parameters, serving for robustness
+            guarantee to avoid one parameter being loaded for multiple times.
+
+        loaded_torch_bins : Set[str]
+            The set of torch binary filenames, serving for robustness guarantee
+            to avoid one torch binary file being loaded for multiple times.
+
+        cached_relax_params : Dict[int, tvm.nd.NDArray]
+            The set of cached Relax parameters.
+
+        cached_torch_params: Dict[str, Any]
+            The set of cached torch parameters. `Any` here stands for numpy.ndarray.
+
+        device : Device
+            The device which we load the parameters to.
+
+        device_cpu : Device
+            The CPU device.
+        """
+        import torch  # pylint: disable=import-outside-toplevel
+
+        assert self.f_convert_pname_fwd is not None
+        assert self.f_convert_param_bkwd is not None
+        assert self.f_compute_relax_param is not None
+        pname2pidx: Dict[str, int] = {
+            pname: pidx for pidx, pname in self.pidx2pname.items()
+        }
+
+        def fetch_torch_param(torch_param):
+            if str(torch_param.dtype) == "torch.bfloat16":
+                # Convert to float32 first.
+                return torch_param.detach().cpu().float().numpy()
+            else:
+                return torch_param.detach().cpu().numpy()
```

Comment:
Hi @vvchernov, I kind of reverted your changes in https://github.com/mlc-ai/mlc-llm/commit/9c3f12c89474d38366d91d61658be615f509654a here. This is mainly because in the standardized parameter loading flow here, we load the raw torch weights without being aware of the concrete quantization method we use.

Regarding the issue you encountered in #458, I wonder if it is possible to address it from another approach: we can specify the dtype conversion of the loaded pytorch parameters similar to the examples below:
https://github.com/mlc-ai/mlc-llm/blob/50b51e0fb8463246d7d39412f9fee9ab871740c6/mlc_llm/relax_model/llama.py#L806-L808
https://github.com/mlc-ai/mlc-llm/blob/50b51e0fb8463246d7d39412f9fee9ab871740c6/mlc_llm/relax_model/rwkv.py#L542-L548

Please let me know if you have further questions! Thank you :-)